keps/sig-api-machinery/storage-migration.md

```diff
@@ -0,0 +1,291 @@
+
+## Motivation
+
+"Today it is possible to create API objects (e.g., HPAs) in one version of
+Kubernetes, go through multiple upgrade cycles without touching those objects,
+and eventually arrive at a version of Kubernetes that can’t interpret the stored
+resource and crashes. See k8s.io/pr/52185."[1][]. We propose a tool that
+prevents this problem.
+
+[1]:https://docs.google.com/document/d/1eoS1K40HLMl4zUyw5pnC05dEF3mzFLp5TPEEt4PFvsM
+
+### Goals
+
+A successful storage version migration tool must:
+* work for Kubernetes built-in APIs, custom resources (CR), and aggregated APIs.
+* do not add burden to cluster administrators or Kubernetes distributions.
+* only cause insignificant load to apiservers. For example, if the master has
+  10GB memory, the migration tool should generate less than 10 qps of single
+  object operations(TODO: measure the memory consumption of PUT operations;
+  study how well the default 10 Mbps bandwidth limit in the oc command work).
+* work for big clusters that have ~10^5 instances of some resource types.
+* make progress in flaky environment, e.g., flaky apiservers, or the migration
+  process get preempted.
+* allow system administrators to track the migration progress.
+
+As to the deliverables,
+* in the short term, providing system administrators with a tool to migrate
+  the API objects to the proper storage versions.
+* in the long term, automating the migration of Kubernetes built-in APIs, CR,
+  aggregated APIs without further burdening system administrators or Kubernetes
+  distributions.
+
+## Proposal
+
+### Workflow (alpha)
+
+We will create a "kube-migrator" command line tool. At the alpha stage,
+"kube-migrator" needs to be manually invoked, and does not handle custom
+resources. 
+
+After all the apiservers are at the desired version, including kube-apiservers
+and aggregated apiservers, the cluster administrator launches "kube-migrator",
+it:
+* deletes any existing `deployment` of "kube-migrator controller". 
+* creates a `deployment` of "kube-migrator controller". 
+* uses the same kubeconfig file as kubectl does.
+
+The "kube-migrator controller" does the following:
+* creates a `progress` custom resource, in the `migration.k8s.io` group.
+  The `progress` has an ownerReference set to the "kube-migrator" deployment.
+* uses the discovery client to discover the supported APIs.
+* lists all custom resource definitions.
+* stores the discovered resource types in the `progress` CR, excluding the
+  custom resources and aggregated APIs.
+* launches jobs, the "migration workers", to migrate each resource type, one by
+  one. Resource types that have only one version are ignored.
+* when a job completes, marks completion in the "progress" CR and deletes the
+  job.
+
+The schema of `progress`:
+```golang
+type Progress struct {
+  metav1.TypeMeta
+  metav1.ObjectMeta
+  // A list of all the resources as suggested by the apiservers
+  Resources []Resource
```

Comment:
The multiplicity here seems odd.  I would expect to have one instance per resource with both spec and status.

Also, shall we split the API and design for the controller out of this first pass?  I accept that such a controller can be built, but I don't see a reason to block the creation of the command line tool on the shape of the controller to come.

Reply:
I think the API should be drivable by an external actor.  I should be able to indicate, "migrate storage for this resource".  That would bump a generation and then the actual migration would reflect a status generation to indicate that migration was complete.  Given an externally driveable change, I suspect you'd also want to be able indicate why you bumped it.

The granularity of both desired and actual state is at the level of resource.  Combining them makes a singleton resource which is generally odd, makes conflicts more likely, and generally means "bad" updates are more likely.

Reply:
I agree with @deads2k about no singleton resources. The "drivable by an external actor" part implies further creation of such an actor. It makes each part of the system more focused and hence composable, at the cost of more moving pieces. Overall I think it's probably slightly better to do it that way.